src/main/java/eu/dissco/core/digitalspecimenprocessor/service/DigitalMediaService.java

```diff
@@ -0,0 +1,136 @@
+import eu.dissco.core.digitalspecimenprocessor.domain.media.DigitalMediaKey;
+import eu.dissco.core.digitalspecimenprocessor.domain.media.DigitalMediaProcessResult;
+import eu.dissco.core.digitalspecimenprocessor.domain.specimen.DigitalSpecimenRecord;
+import eu.dissco.core.digitalspecimenprocessor.domain.specimen.UpdatedDigitalSpecimenRecord;
+import eu.dissco.core.digitalspecimenprocessor.repository.DigitalMediaRepository;
+import eu.dissco.core.digitalspecimenprocessor.schema.EntityRelationship;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+import java.util.stream.Collectors;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+
+@Service
+@Slf4j
+@RequiredArgsConstructor
+public class DigitalMediaService {
+
+  private final DigitalMediaRepository mediaRepository;
+
+  public Map<String, DigitalMediaProcessResult> getExistingDigitalMedia(
+      Map<String, DigitalSpecimenRecord> currentSpecimens,
+      List<DigitalMediaEventWithoutDOI> mediaEvents) {
+    var mediaProcessResults = new HashMap<String, DigitalMediaProcessResult>();
+    var mediaMap = new HashMap<String, List<DigitalMediaEventWithoutDOI>>();
+    mediaEvents.forEach(mediaEvent ->
+        mediaMap.computeIfAbsent(mediaEvent.digitalMediaObjectWithoutDoi().physicalSpecimenID(),
+            k -> new ArrayList<>()).add(mediaEvent)
+    );
+    var mediaIds = getAllMediaIds(
+        currentSpecimens.values()); //DigitalMediaKey (specimenId, mediaUrl), mediaId
+    for (var entry : currentSpecimens.entrySet()) {
+      var currentSpecimen = entry.getValue();
+      // use physical specimen id to get relevant media events
+      var mediaEventsForSpecimen =
+          mediaMap.get(entry.getKey()) == null ?
+              new ArrayList<DigitalMediaEventWithoutDOI>() : mediaMap.get(entry.getKey());
+      var existingMediaForCurrentSpecimen = mediaIds.entrySet().stream()
+          .filter(e -> e.getKey().digitalSpecimenId().equals(currentSpecimen.id()))
+          .collect(Collectors.toMap(Entry::getKey, Entry::getValue));
+      var processResult = getExistingDigitalMediaProcessResult(
+          currentSpecimen.digitalSpecimenWrapper().attributes().getOdsHasEntityRelationship(),
+          mediaEventsForSpecimen, existingMediaForCurrentSpecimen, currentSpecimen.id());
+      mediaProcessResults.put(currentSpecimen.id(), processResult);
+    }
+    return mediaProcessResults;
+  }
+
+
+  /*
+   In our (incoming) mediaEvents, we are missing mediaID but we have the url
+   In the existing entity relationships in the specimen, we are missing the media url but we have the id
+   This function gets the mediaUrls based on the media id from the ER
+   That way we can match incoming media events to existing media objects and find which media are new
+   */
+
```

Comment:
Not sure is it 'normal' to add a white line here. I would say add it directly above.